### PR TITLE
Dynamo

### PR DESCRIPTION
**Pull request by jeteve**
This implements a first version of the DynamoDB storage.

### CODE REVIEW

File: cpanfile [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
 requires 'Log::Any' , '>= 1.032';
 requires 'Moose' , '>= 2.1405';
 requires 'Class::Load';
 requires 'Scope::Guard', '>= 0.21';
 
 test_requires 'Test::More';
-test_requires 'Test::MockDateTime';
\ No newline at end of file
+test_requires 'Test::MockDateTime';
+test_requires 'Class::Load';
```
Comment (stuartskelton) [inline cpanfile]: is this necessary, as it is also a dep of the main library?

File: t/storage-dynamodb.t [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                   'Paws::Net::LWPCaller',
+                   'JSON',
+                   'MIME::Base64',
+                   'Compress::Zlib',
+               );
+
+join( '', map{ Class::Load::try_load_class( $_ ) ? 'yes' : '' } @paws_class ) eq join('', map{ 'yes' } @paws_class )
+    or plan skip_all => "Paws required to run these tests";
+
+$ENV{DYNAMODB_LOCAL} or plan skip_all => "ENV DYNAMODB_LOCAL URL is required";
```
Comment (stuartskelton) [inline t/storage-dynamodb.t]: should  this not be a little higher up as there is no point loading modules if you are going to skip

File: lib/Schedule/LongSteps/Storage/DynamoDB.pm [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+  my $sl = Schedule::LongSteps->new({ storage => $storage, ... }, ... );
+
+=cut
+
+has 'dynamo_db' => (is => 'ro', isa => 'Paws::DynamoDB' , required => 1 );
+has 'table_prefix' => ( is => 'ro', isa => 'Str', required => 1);
+
+has 'table_name' => ( is => 'ro', isa => 'Str', lazy_build => 1);
+
+has 'creation_wait_time' => ( is => 'ro', isa => 'Int', default => 2 );
```
Comment (stuartskelton) [inline lib/Schedule/LongSteps/Storage/DynamoDB.pm:42]: used for waiting on the creation and deletion of the table.

File: lib/Schedule/LongSteps/Storage/DynamoDB.pm [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+    my @found_items = ();
+    my $query_output;
+    do{
+        # Query the by_run_at_day index
+        $query_output = $self->dynamo_db()->Query(
+            TableName => $self->table_name(),
+            IndexName => 'by_run_at_day',
+            ConsistentRead => 0, # Consistent read are not supported on secondary indices.
+            ExpressionAttributeValues => {
+                ":run_at_day" => { "S" => substr( $run_at_day->iso8601() , 0 , 10 ) },
```
Comment (stuartskelton) [inline lib/Schedule/LongSteps/Storage/DynamoDB.pm:201]: would it not be safer (and a little more readable) if you used '$dt->strftime'?

File: lib/Schedule/LongSteps/Storage/DynamoDB.pm [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+    return $self;
+}
+
+sub _state_encode{
+    my ($self) = @_;
+    my $state_json = JSON::to_json(
+        $self->state(),
+        { ascii => 1 }
+    );
+    unless( length( $state_json ) > 350_000 ){
```
Comment (stuartskelton) [inline lib/Schedule/LongSteps/Storage/DynamoDB.pm]: should this magic number not be a named constant?

File: lib/Schedule/LongSteps/Storage/DynamoDB.pm [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+                    ":null" => { "S" => 'NULL' },
+                },
+                Key => { id => { S => $id } },
+                UpdateExpression => 'SET run_id = :run_id',
+                ConditionExpression => 'run_id = :null',
+                ReturnValues => 'ALL_NEW',
+            );
+        };
+        if( my $err = $@ ){
+            # This error is expected in case of race condition.
```
Comment (stuartskelton) [inline lib/Schedule/LongSteps/Storage/DynamoDB.pm:243]: such recovery

File: lib/Schedule/LongSteps/Storage/DynamoDB.pm [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+
+    $log->info("Inserting ".ref($self)." , id = ".$self->id()." in Dynamo table ".$table_name);
+    $self->storage()->dynamo_db->PutItem(
+        TableName => $table_name,
+        Item => $self->_to_dynamo_item(),
+    );
+    return $self;
+}
+
+sub _state_encode{
```
Comment (stuartskelton) [inline lib/Schedule/LongSteps/Storage/DynamoDB.pm:487]: _state_encode _decode_state, might be worth, fixing the convention (and maybe moving them closer together

File: lib/Schedule/LongSteps/Storage/DynamoDB.pm [hunk 8]
```diff
@@ -0,0 +1,538 @@
+package Schedule::LongSteps::Storage::DynamoDB;
+
+use Moose;
+extends qw/Schedule::LongSteps::Storage/;
+
+use Log::Any qw/$log/;
+use Scalar::Util;
+use MIME::Base64;
+use Compress::Zlib;
+use JSON;
```
Comment (stuartskelton) [inline lib/Schedule/LongSteps/Storage/DynamoDB.pm]: use JSON (); ?

File: lib/Schedule/LongSteps/Storage/DynamoDB.pm [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+use Compress::Zlib;
+use DateTime;
+use DateTime::Format::ISO8601;
+use Data::Dumper;
+
+has 'storage' => ( is => 'ro', isa => 'Schedule::LongSteps::Storage::DynamoDB', required => 1 );
+has 'id' =>      ( is => 'ro', isa => 'Str', required => 1 );
+
+has 'process_class' => ( is => 'rw', isa => 'Str', required => 1); # rw only for test. Should not changed ever.
+has 'status' =>        ( is => 'rw', isa => 'Str', default => 'pending' );
```
Comment (stuartskelton) [inline lib/Schedule/LongSteps/Storage/DynamoDB.pm:397]: This seems a little messy, when you have taken the time to align the "is  => 'rw'"

File: lib/Schedule/LongSteps/Storage/DynamoDB.pm (hunk 4 above)
Comment (jeteve) [reply lib/Schedule/LongSteps/Storage/DynamoDB.pm:201]: could use ymd actually :)

**Pull request merged**